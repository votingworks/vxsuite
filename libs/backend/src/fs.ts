import { Result, assert, err, ok } from '@votingworks/basics';
import { Buffer } from 'buffer';
import * as fs from 'fs';
import { promisify } from 'util';

const fsReadFile = promisify(fs.readFile);

/**
 * The number of bytes in a megabyte.
 */
export const ONE_MEGABYTE = 1024 * 1024;

/**
 *
 */
export type ReadFileError =
  | { type: 'file-not-found'; path: string }
  | { type: 'file-too-large'; path: string; maxSize: number };

/**
 * The result of reading a file from the file system.
 */
export type ReadFileResult<T extends string | Buffer> = Result<
  T,
  ReadFileError
>;

/**
 * Reads a file from the file system up to a maximum size.
 *
 * @param path The path to the file to read.
 * @param options.maxSize The maximum size of the file to read.
 * @param options.encoding The encoding to use when reading the file, if any.
 * @returns A promise that resolves to the contents of the file.
 */
export async function readFile(
  path: string,
  options: { maxSize: number }
): Promise<ReadFileResult<Buffer>>;

/**
 * Reads a file from the file system up to a maximum size.
 *
 * @param path The path to the file to read.
 * @param options.maxSize The maximum size of the file to read.
 * @param options.encoding The encoding to use when reading the file, if any.
 * @returns A promise that resolves to the contents of the file.
 */
export async function readFile(
  path: string,
  options: { maxSize: number; encoding: BufferEncoding }
): Promise<ReadFileResult<string>>;

/**
 * Reads a file from the file system up to a maximum size.
 *
 * @param path The path to the file to read.
 * @param options.maxSize The maximum size of the file to read.
 * @param options.encoding The encoding to use when reading the file, if any.
 * @returns A promise that resolves to the contents of the file.
 */
export async function readFile(
  path: string,
  options: { maxSize: number; encoding?: BufferEncoding }
): Promise<ReadFileResult<string | Buffer>> {
  const { maxSize = ONE_MEGABYTE, encoding } = options ?? {};
  assert(maxSize > 0, 'maxSize must be greater than zero');
  assert(
    Number.isInteger(maxSize) || !Number.isFinite(maxSize),
    'maxSize must be an integer or Infinity'
  );

  if (!Number.isFinite(maxSize)) {
    return ok(await fsReadFile(path, { encoding }));
  }

  const input = fs.createReadStream(path, { encoding });

  return new Promise((resolve, reject) => {
    let bytesRead = 0;
    const chunks: unknown[] = [];

    input.on('readable', () => {
      // read one more byte than we need to find out if the file is too large
      const chunk = input.read(maxSize - bytesRead + 1);

      // if we've reached the end of the file, we're done
      if (chunk === null) {
        return;
      }

      if (Buffer.isBuffer(chunk)) {
        bytesRead += chunk.length;
      } else if (typeof chunk === 'string') {
        bytesRead += Buffer.byteLength(chunk);
      } else {
        throw new Error('Unexpected chunk type');
      }

      if (bytesRead > maxSize) {
        input.destroy();
        resolve(
          err({
            type: 'file-too-large',
            path,
            maxSize,
          })
        );
        return;
      }

      chunks.push(chunk);
    });

    input.on('error', (error) => {
      if ((error as { code?: string }).code === 'ENOENT') {
        resolve(
          err({
            type: 'file-not-found',
            path,
          })
        );
      } else {
        reject(error);
      }
    });

    input.on('end', () => {
      resolve(
        ok(encoding ? chunks.join('') : Buffer.concat(chunks as Buffer[]))
      );
    });
  });
}
