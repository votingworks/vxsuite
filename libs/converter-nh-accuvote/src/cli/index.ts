/**
 * Represents IO for a CLI. Facilitates mocking for testing.
 */
export interface Stdio {
  readonly stdin: NodeJS.ReadableStream;
  readonly stdout: NodeJS.WritableStream;
  readonly stderr: NodeJS.WritableStream;
}

/**
 * The default IO implementation.
 */
export const RealIo: Stdio = {
  stdin: process.stdin,
  stdout: process.stdout,
  stderr: process.stderr,
};
