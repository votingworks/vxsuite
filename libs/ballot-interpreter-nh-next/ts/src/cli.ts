import { find, groupBy } from '@votingworks/basics';
import {
  ElectionDefinition,
  Optional,
  safeParseElectionDefinition,
  SheetOf,
} from '@votingworks/types';
import { jsonStream } from '@votingworks/utils';
import Sqlite3 from 'better-sqlite3';
import chalk from 'chalk';
import { promises as fs } from 'fs';
import { join } from 'path';
import { pipeline } from 'stream/promises';
import { interpret } from './interpret';
import { InterpretedBallotCard, InterpretError } from './types';

function usage(out: NodeJS.WritableStream): void {
  out.write(
    `Usage: interpret <election.json> <ballot-side-a.jpeg> <ballot-side-b.jpeg>\n` +
      `       interpret <scan-workspace> [<sheet-id> …]\n`
  );
}

function prettyPrintInterpretation({
  electionDefinition,
  paths,
  stdout,
  interpretedBallotCard,
}: {
  electionDefinition: ElectionDefinition;
  paths: SheetOf<string>;
  stdout: NodeJS.WritableStream;
  interpretedBallotCard: InterpretedBallotCard;
}) {
  const thresholds = electionDefinition.election.markThresholds;

  if (!thresholds) {
    stdout.write(
      `Warning: No mark thresholds defined in election definition; cannot render marks as votes.\n`
    );
  }

  stdout.write(`${chalk.bold('Paths:')}\n`);
  stdout.write(`- ${paths[0]}\n`);
  stdout.write(`- ${paths[1]}\n`);
  stdout.write(`\n`);

  const marksByContest = groupBy(
    [...interpretedBallotCard.front.marks, ...interpretedBallotCard.back.marks],
    ([{ contestId }]) => contestId
  );

  for (const [contestId, marks] of marksByContest) {
    const contest = find(
      electionDefinition.election.contests,
      (c) => c.id === contestId
    );
    stdout.write(`${chalk.italic(contest.title)}\n`);

    for (const [gridPosition, scoredMark] of marks) {
      const candidate =
        contest.type === 'candidate' && gridPosition.type === 'option'
          ? contest.candidates.find((c) => c.id === gridPosition.optionId)
          : undefined;
      const displayName =
        contest.type === 'candidate'
          ? gridPosition.type === 'option'
            ? candidate?.name ?? gridPosition.optionId
            : `Write-In #${gridPosition.writeInIndex + 1}`
          : contest.type === 'yesno'
          ? gridPosition.type === 'option' && gridPosition.optionId === 'yes'
            ? contest.yesOption?.label ?? 'Yes'
            : contest.noOption?.label ?? 'No'
          : 'Unknown';

      stdout.write(
        `${
          !scoredMark || !thresholds
            ? ' '
            : scoredMark.fillScore < thresholds.marginal
            ? '⬜️'
            : scoredMark.fillScore < thresholds.definite
            ? '❓'
            : '✅'
        } ${
          scoredMark
            ? chalk.dim(
                `(${(scoredMark.fillScore * 100).toFixed(2).padStart(5)}%)`
              )
            : ''
        } ${displayName}\n`
      );
    }

    stdout.write('\n');
  }
}

async function interpretFiles(
  electionDefinitionOrPath: ElectionDefinition | string,
  [ballotPathSideA, ballotPathSideB]: SheetOf<string>,
  {
    stdout,
    stderr,
    json = false,
    debug = false,
  }: {
    stdout: NodeJS.WritableStream;
    stderr: NodeJS.WritableStream;
    json?: boolean;
    debug?: boolean;
  }
): Promise<number> {
  let electionDefinition: ElectionDefinition;

  if (typeof electionDefinitionOrPath === 'string') {
    const parseElectionDefinitionResult = safeParseElectionDefinition(
      await fs.readFile(electionDefinitionOrPath, 'utf8')
    );

    if (parseElectionDefinitionResult.isErr()) {
      stderr.write(
        `Error parsing election definition: ${
          parseElectionDefinitionResult.err().message
        }\n`
      );
      usage(stderr);
      return 1;
    }

    electionDefinition = parseElectionDefinitionResult.ok();
  } else {
    electionDefinition = electionDefinitionOrPath;
  }

  const result = interpret(
    electionDefinition,
    [ballotPathSideA, ballotPathSideB],
    { debug }
  );

  if (result.isErr()) {
    stderr.write(`Error interpreting ballot:\n`);
    await pipeline(jsonStream<InterpretError>(result.err()), stderr);
    return 1;
  }

  if (json) {
    await pipeline(jsonStream(result.ok(), { compact: false }), stdout);
  } else {
    prettyPrintInterpretation({
      electionDefinition,
      paths: [ballotPathSideA, ballotPathSideB],
      interpretedBallotCard: result.ok(),
      stdout,
    });
  }

  return 0;
}

async function interpretWorkspace(
  workspacePath: string,
  {
    stdout,
    stderr,
    sheetIds,
    json = false,
    debug = false,
  }: {
    stdout: NodeJS.WriteStream;
    stderr: NodeJS.WriteStream;
    sheetIds: Iterable<string>;
    json?: boolean;
    debug?: boolean;
  }
): Promise<number> {
  const dbPath = join(workspacePath, 'ballots.db');

  if (!(await fs.stat(dbPath)).isFile()) {
    stderr.write(`No ballots database found in workspace.\n`);
    stderr.write(`Expected to find ${dbPath}.\n`);
    return 1;
  }

  const db = new Sqlite3(dbPath);
  const { electionData } =
    (db
      .prepare('SELECT election_data as electionData FROM election LIMIT 1')
      .get() as Optional<{ electionData: string }>) ?? {};

  if (!electionData) {
    stderr.write(
      `No election data found in workspace. Is the workspace configured?\n`
    );
    return 1;
  }

  const electionDefinition =
    safeParseElectionDefinition(electionData).unsafeUnwrap();

  const sheetIdsArray = [...sheetIds];
  const sheets = (
    sheetIdsArray.length
      ? db
          .prepare(
            'SELECT id, front_original_filename as frontPath, back_original_filename as backPath FROM sheets WHERE id IN ?'
          )
          .all(sheetIdsArray)
      : db
          .prepare(
            'SELECT id, front_original_filename as frontPath, back_original_filename as backPath FROM sheets'
          )
          .all()
  ) as Array<{ id: string; frontPath: string; backPath: string }>;

  for (const { id, frontPath, backPath } of sheets) {
    stdout.write(`${chalk.bold('Sheet ID:')} ${id}\n`);
    await interpretFiles(electionDefinition, [frontPath, backPath], {
      stdout,
      stderr,
      json,
      debug,
    });
  }

  return 0;
}

/**
 * CLI for running the interpreter standalone.
 */
export async function main(args: string[]): Promise<number> {
  const { stdout, stderr } = process;
  let workspacePath: string | undefined;
  const sheetIds = new Set<string>();
  let electionDefinitionPath: string | undefined;
  let ballotPathSideA: string | undefined;
  let ballotPathSideB: string | undefined;
  let json = false;
  let debug = false;

  for (const arg of args) {
    if (arg === '-h' || arg === '--help') {
      usage(stdout);
      return 0;
    }

    if (arg === '-j' || arg === '--json') {
      json = true;
      continue;
    }

    if (arg === '-d' || arg === '--debug') {
      debug = true;
      continue;
    }

    if (arg.startsWith('-')) {
      stderr.write(`Unknown option: ${arg}\n`);
      usage(stderr);
      return 1;
    }

    if (!workspacePath && !electionDefinitionPath) {
      const stat = await fs.stat(arg);
      if (stat.isDirectory()) {
        workspacePath = arg;
      } else if (stat.isFile()) {
        electionDefinitionPath = arg;
      } else {
        stderr.write(
          `Expected an election definition file or scan workspace: ${arg}\n`
        );
        usage(stderr);
        return 1;
      }
    } else if (electionDefinitionPath && !ballotPathSideA) {
      ballotPathSideA = arg;
    } else if (electionDefinitionPath && !ballotPathSideB) {
      ballotPathSideB = arg;
    } else if (workspacePath) {
      sheetIds.add(arg);
    } else {
      stderr.write(`Unexpected argument: ${arg}\n`);
      usage(stderr);
      return 1;
    }
  }

  if (workspacePath) {
    return await interpretWorkspace(workspacePath, {
      sheetIds,
      stdout,
      stderr,
      json,
      debug,
    });
  }

  if (electionDefinitionPath && ballotPathSideA && ballotPathSideB) {
    return await interpretFiles(
      electionDefinitionPath,
      [ballotPathSideA, ballotPathSideB],
      { stdout, stderr, json, debug }
    );
  }

  usage(stderr);
  return 1;
}
