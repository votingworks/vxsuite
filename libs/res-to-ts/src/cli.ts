import * as fs from 'fs/promises';
import globby from 'globby';
import { basename, dirname, isAbsolute, join, relative } from 'path';
import { convert, Resource } from './convert';
import { getMimeType } from './mime';
import { assert } from './utils/assert';

/**
 * Represents IO for a CLI. Facilitates mocking for testing.
 */
export interface Stdio {
  readonly stdin: NodeJS.ReadableStream;
  readonly stdout: NodeJS.WritableStream;
  readonly stderr: NodeJS.WritableStream;
}

/**
 * The default IO implementation.
 */
export const RealIo: Stdio = {
  stdin: process.stdin,
  stdout: process.stdout,
  stderr: process.stderr,
};

/**
 * All possible options for the CLI.
 */
export type Options = HelpOptions | ConvertOptions;

/**
 * Options for the CLI when run with `--help`.
 */
export interface HelpOptions {
  readonly type: 'help';
  readonly help: true;
}

/**
 * Options for the CLI's default mode.
 */
export interface ConvertOptions {
  readonly type: 'convert';
  readonly help: false;
  readonly check: boolean;
  readonly rootDir?: string;
  readonly outDir?: string;
  readonly resources: readonly Resource[];
}

/**
 * Result of parsing command-line arguments.
 */
type OptionParseResult = Options | OptionParseError;

/**
 * Errors returned by {@link parseOptions}.
 */
interface OptionParseError {
  type: 'error';
  error: Error;
}

/**
 * Prints command usage to {@link out}.
 */
function usage(programName: string, out: NodeJS.WritableStream): void {
  out.write(`usage: ${programName} [--help] [--check] FILE [‚Ä¶ FILE]\n`);
  out.write(`\n`);
  out.write(`Converts resources to be usable as TypeScript files.\n`);
}

/**
 * @visibleForTesting
 */
export function absolutize(path: string, base: string): string {
  return isAbsolute(path) ? path : join(base, path);
}

/**
 * @visibleForTesting
 */
export function relativize(path: string, base?: string): string {
  return base ? relative(base, path) : path;
}

/**
 * Determines the output `.ts` path for a resource.
 */
export function getOutputPath(
  resourcePath: string,
  rootDir?: string,
  outDir?: string
): string {
  if (!outDir) {
    return `${resourcePath}.ts`;
  }

  assert(typeof rootDir === 'string');

  const relativePath = relative(rootDir, resourcePath);
  return `${join(outDir, relativePath)}.ts`;
}

/**
 * Parses and validates command-line arguments.
 */
async function parseOptions(
  args: readonly string[]
): Promise<OptionParseResult> {
  let help = false;
  let check = false;
  let rootDir: string | undefined;
  let outDir: string | undefined;
  const globs: string[] = [];
  const resources: Resource[] = [];

  for (let i = 2; i < args.length; i += 1) {
    const arg = args[i] as string;
    switch (arg) {
      case '-h':
      case '--help':
        help = true;
        break;

      case '--check':
        check = true;
        break;

      case '-o':
      case '--outDir': {
        outDir = args[i + 1];
        if (!outDir || outDir.startsWith('-')) {
          return {
            type: 'error',
            error: new Error(`missing output directory after '${arg}'`),
          };
        }
        outDir = absolutize(outDir, process.cwd());
        i += 1;
        break;
      }

      case '--rootDir': {
        rootDir = args[i + 1];
        if (!rootDir || rootDir.startsWith('-')) {
          return {
            type: 'error',
            error: new Error(`missing root directory after '${arg}'`),
          };
        }
        rootDir = absolutize(rootDir, process.cwd());
        i += 1;
        break;
      }

      default: {
        const glob = arg;
        if (glob.startsWith('-')) {
          return {
            type: 'error',
            error: new Error(`unrecognized option: ${glob}`),
          };
        }
        globs.push(glob);
        break;
      }
    }
  }

  if (help) {
    return { type: 'help', help };
  }

  if (outDir && !rootDir) {
    rootDir = process.cwd();
  }

  for (const path of await globby(globs, {
    absolute: true,
    onlyFiles: false,
  })) {
    if (rootDir && !path.startsWith(`${rootDir}/`)) {
      return {
        type: 'error',
        error: new Error(
          `resource '${path}' is not in the root directory '${rootDir}'`
        ),
      };
    }

    resources.push({
      path,
      tsPath: getOutputPath(path, rootDir, outDir),
      mimeType: getMimeType(path),
    });
  }

  if (resources.length === 0) {
    return { type: 'error', error: new Error('no resources given') };
  }

  return { type: 'convert', help, check, rootDir, outDir, resources };
}

/**
 * Main entry point for `res-to-ts` command.
 */
export async function main(
  args: readonly string[],
  io: Stdio
): Promise<number> {
  const programName = basename(args[1] as string);
  const parseOptionsResult = await parseOptions(args);

  if (parseOptionsResult.type === 'error') {
    io.stderr.write(`error: ${parseOptionsResult.error.message}\n`);
    usage(programName, io.stderr);
    return 1;
  }

  const options = parseOptionsResult;

  if (options.help) {
    usage(programName, io.stdout);
    return 0;
  }

  const { check, resources } = options;
  let exitCode = 0;

  for (const resource of resources) {
    const output = await convert(resource);
    const displayPath = relativize(resource.tsPath, process.cwd());

    if (check) {
      let success = true;
      try {
        const existingContent = await fs.readFile(resource.tsPath, 'utf8');
        success = existingContent === output;
      } catch {
        success = false;
      }

      if (!success) {
        io.stderr.write(`‚ùå ${displayPath} is out of date\n`);
        exitCode = 1;
      } else {
        io.stdout.write(`‚úÖ ${displayPath}\n`);
      }
    } else {
      await fs.mkdir(dirname(resource.tsPath), { recursive: true });
      await fs.writeFile(resource.tsPath, output);
      io.stdout.write(`üìù ${displayPath}\n`);
    }
  }

  return exitCode;
}
