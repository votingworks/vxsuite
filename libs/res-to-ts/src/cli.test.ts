import { fakeReadable, fakeWritable } from '@votingworks/test-utils';
import { dirSync, fileSync } from 'tmp';
import { promises as fs } from 'fs';
import { join } from 'path';
import { main } from '.';
import { absolutize, getOutputPath, relativize, Stdio } from './cli';

test('absolutize', () => {
  expect(absolutize('/foo/bar', '/baz')).toEqual('/foo/bar');
  expect(absolutize('foo/bar', '/baz')).toEqual('/baz/foo/bar');
});

test('relativize', () => {
  expect(relativize('/foo/bar', '/foo')).toEqual('bar');
  expect(relativize('/foo/bar', '/baz')).toEqual('../foo/bar');
  expect(relativize('/foo/bar')).toEqual('/foo/bar');
});

test('-h', async () => {
  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(['node', '/path/to/res-to-ts', '-h'], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toMatchInlineSnapshot(`
    Object {
      "exitCode": 0,
      "stderr": "",
      "stdout": "usage: res-to-ts [--help] [--check] FILE [… FILE]

    Converts resources to be usable as TypeScript files.
    ",
    }
  `);
});

test('--help', async () => {
  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(['node', '/path/to/res-to-ts', '--help'], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toMatchInlineSnapshot(`
    Object {
      "exitCode": 0,
      "stderr": "",
      "stdout": "usage: res-to-ts [--help] [--check] FILE [… FILE]

    Converts resources to be usable as TypeScript files.
    ",
    }
  `);
});

test('invalid option', async () => {
  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', '--invalid'],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toMatchInlineSnapshot(`
    Object {
      "exitCode": 1,
      "stderr": "error: unrecognized option: --invalid
    usage: res-to-ts [--help] [--check] FILE [… FILE]

    Converts resources to be usable as TypeScript files.
    ",
      "stdout": "",
    }
  `);
});

test('missing rootDir', async () => {
  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', '--rootDir'],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toMatchInlineSnapshot(`
    Object {
      "exitCode": 1,
      "stderr": "error: missing root directory after '--rootDir'
    usage: res-to-ts [--help] [--check] FILE [… FILE]

    Converts resources to be usable as TypeScript files.
    ",
      "stdout": "",
    }
  `);
});

test('missing outDir', async () => {
  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', '--outDir'],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toMatchInlineSnapshot(`
    Object {
      "exitCode": 1,
      "stderr": "error: missing output directory after '--outDir'
    usage: res-to-ts [--help] [--check] FILE [… FILE]

    Converts resources to be usable as TypeScript files.
    ",
      "stdout": "",
    }
  `);
});

test('no files given', async () => {
  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(['node', '/path/to/res-to-ts'], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 1,
    stdout: '',
    stderr: expect.stringContaining('error: no resources given'),
  });
});

test('convert image adjacent', async () => {
  const path = fileSync({ template: 'XXXXXX.png' }).name;
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const tsPath = getOutputPath(path);
  const exitCode = await main(['node', '/path/to/res-to-ts', path], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: `📝 ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });

  const content = await fs.readFile(tsPath, 'utf8');
  expect(content).toContain(`const resourceDataBase64 = 'Y29udGVudA=='`);
  expect(content).toContain(`const mimeType = 'image/png'`);
  expect(content).toContain(`export async function asImage()`);
});

test('convert w/outDir', async () => {
  const root = dirSync().name;
  const rootDir = join(root, 'input');
  const outDir = join(root, 'output');
  const path = join(rootDir, 'image.png');
  await fs.mkdir(rootDir, { recursive: true });
  await fs.mkdir(outDir, { recursive: true });
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const tsPath = getOutputPath(path, rootDir, outDir);
  const exitCode = await main(
    [
      'node',
      '/path/to/res-to-ts',
      path,
      '--rootDir',
      rootDir,
      '--outDir',
      outDir,
    ],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: `📝 ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });

  const content = await fs.readFile(tsPath, 'utf8');
  expect(content).toContain(`const resourceDataBase64 = 'Y29udGVudA=='`);
  expect(content).toContain(`const mimeType = 'image/png'`);
  expect(content).toContain(`export async function asImage()`);
});

test('convert w/implicit rootDir', async () => {
  const root = dirSync().name;
  const rootDir = join(root, 'input');
  const outDir = join(root, 'output');
  const path = join(rootDir, 'image.png');
  await fs.mkdir(rootDir, { recursive: true });
  await fs.mkdir(outDir, { recursive: true });
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const oldPwd = process.cwd();
  try {
    process.chdir(rootDir);
    const tsPath = getOutputPath(path, rootDir, outDir);
    const exitCode = await main(
      ['node', '/path/to/res-to-ts', path, '--outDir', outDir],
      stdio
    );
    expect({
      exitCode,
      stdout: stdio.stdout.toString(),
      stderr: stdio.stderr.toString(),
    }).toEqual({
      exitCode: 0,
      stdout: `📝 ${relativize(tsPath, rootDir)}\n`,
      stderr: '',
    });

    const content = await fs.readFile(tsPath, 'utf8');
    expect(content).toContain(`const resourceDataBase64 = 'Y29udGVudA=='`);
    expect(content).toContain(`const mimeType = 'image/png'`);
    expect(content).toContain(`export async function asImage()`);
  } finally {
    process.chdir(oldPwd);
  }
});

test('convert failure: resource not in rootDir', async () => {
  const path = fileSync({ template: 'XXXXXX.png' }).name;
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', path, '-o', '/path/to/output'],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 1,
    stdout: ``,
    stderr: expect.stringContaining(
      `resource '${path}' is not in the root directory`
    ),
  });
});

test('convert text adjacent', async () => {
  const path = fileSync({ template: 'XXXXXX.txt' }).name;
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const tsPath = getOutputPath(path);
  const exitCode = await main(['node', '/path/to/res-to-ts', path], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: `📝 ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });

  const content = await fs.readFile(tsPath, 'utf8');
  expect(content).toContain(`const resourceDataBase64 = 'Y29udGVudA=='`);
  expect(content).toContain(`const mimeType = 'text/plain'`);
  expect(content).toContain(`export function asText()`);
  expect(content).not.toContain(`export async function asImage()`);
});

test('convert Election JSON adjacent', async () => {
  const path = fileSync({ template: 'election-XXXXXX.json' }).name;
  await fs.writeFile(path, '{}');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const tsPath = getOutputPath(path);
  const exitCode = await main(['node', '/path/to/res-to-ts', path], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: `📝 ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });

  const content = await fs.readFile(tsPath, 'utf8');
  expect(content).toContain(`const resourceDataBase64 = 'e30='`);
  expect(content).toContain(
    `SHA-256 hash of file data: 44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a`
  );
  expect(content).toContain(`const mimeType = 'application/json'`);
  expect(content).toContain(`export function asText()`);
  expect(content).toContain(`export const electionDefinition =`);
  expect(content).not.toContain(`export async function asImage()`);
});

test('convert unknown type adjacent', async () => {
  const path = fileSync({ template: 'XXXXXX' }).name;
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const tsPath = getOutputPath(path);
  const exitCode = await main(['node', '/path/to/res-to-ts', path], stdio);
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: `📝 ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });

  const content = await fs.readFile(tsPath, 'utf8');
  expect(content).toContain(`const resourceDataBase64 = 'Y29udGVudA=='`);
  expect(content).toContain(`const mimeType = 'application/octet-stream'`);
  expect(content).not.toContain(`export async function asImage()`);
});

test('check success adjacent', async () => {
  const path = fileSync({ template: 'XXXXXX.jpeg' }).name;
  const tsPath = getOutputPath(path);
  await fs.writeFile(path, 'content');

  let stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  await main(['node', '/path/to/res-to-ts', path], stdio);

  expect({
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    stdout: `📝 ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });

  stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', '--check', path],
    stdio
  );

  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: `✅ ${relativize(tsPath, process.cwd())}\n`,
    stderr: '',
  });
});

test('check failure: no .ts file adjacent', async () => {
  const path = fileSync({ template: 'XXXXXX.jpeg' }).name;
  const tsPath = getOutputPath(path);
  await fs.writeFile(path, 'content');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', '--check', path],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 1,
    stderr: `❌ ${relativize(tsPath, process.cwd())} is out of date\n`,
    stdout: '',
  });
});

test('check failure: .ts file outdated adjacent', async () => {
  const path = fileSync({ template: 'XXXXXX.jpeg' }).name;
  const tsPath = getOutputPath(path);
  await fs.writeFile(path, 'content');
  await fs.writeFile(tsPath, '/* outdated */');

  const stdio: Stdio = {
    stdin: fakeReadable(),
    stdout: fakeWritable(),
    stderr: fakeWritable(),
  };

  const exitCode = await main(
    ['node', '/path/to/res-to-ts', '--check', path],
    stdio
  );
  expect({
    exitCode,
    stdout: stdio.stdout.toString(),
    stderr: stdio.stderr.toString(),
  }).toEqual({
    exitCode: 1,
    stderr: `❌ ${relativize(tsPath, process.cwd())} is out of date\n`,
    stdout: '',
  });
});
