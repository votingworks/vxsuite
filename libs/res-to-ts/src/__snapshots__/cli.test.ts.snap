// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`-h 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": 
"usage: res-to-ts [--help] [--check] FILE [… FILE]

Converts resources to be usable as TypeScript files.
"
,
}
`;

exports[`--help 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": 
"usage: res-to-ts [--help] [--check] FILE [… FILE]

Converts resources to be usable as TypeScript files.
"
,
}
`;

exports[`invalid option 1`] = `
{
  "exitCode": 1,
  "stderr": 
"error: unrecognized option: --invalid
usage: res-to-ts [--help] [--check] FILE [… FILE]

Converts resources to be usable as TypeScript files.
"
,
  "stdout": "",
}
`;

exports[`missing rootDir 1`] = `
{
  "exitCode": 1,
  "stderr": 
"error: missing root directory after '--rootDir'
usage: res-to-ts [--help] [--check] FILE [… FILE]

Converts resources to be usable as TypeScript files.
"
,
  "stdout": "",
}
`;

exports[`missing outDir 1`] = `
{
  "exitCode": 1,
  "stderr": 
"error: missing output directory after '--outDir'
usage: res-to-ts [--help] [--check] FILE [… FILE]

Converts resources to be usable as TypeScript files.
"
,
  "stdout": "",
}
`;
